batchproducer/batchproducer.go

```diff
@@ -0,0 +1,314 @@
+package batchproducer
+
+import (
+	"errors"
+	"log"
+	"time"
+
+	"github.com/timehop/go-kinesis"
+)
+
+// BatchProducer collects records individually and then sends them to Kinesis in
+// batches in the background using PutRecords, with retries.
+// A BatchProducer will do nothing until Start is called.
+type BatchProducer interface {
+	// Start starts the main goroutine. No need to call it using `go`.
+	Start() error
+	Stop() error
+
+	// Add might block if the BatchProducer has a buffer and the buffer is full.
+	// In order to prevent filling the buffer and eventually blocking indefinitely,
+	// Add will fail and return an error if the BatchProducer is not started.
+	Add(data []byte, partitionKey string) error
+
+	// Setters
+	SetMaxAttemptsPerRecord(int) error
+	SetStatReceiver(BatchStatReceiver) error
+
+	// This interval will be used to make a *best effort* attempt to send stats *approximately*
+	// when this interval elapses. There’s no guarantee, however, since the main goroutine is
+	// used to send the stats and therefore there may be some skew.
+	SetStatInterval(time.Duration) error
+}
+
+// BatchStatReceiver defines an object that can accept stats.
+type BatchStatReceiver interface {
+	// Receive will be called by the main BatchProducer goroutine so it will block all batches from being
+	// sent, so make sure it is either very fast or never blocks at all!
+	Receive(BatchStat)
+}
+
+// BatchStat is a kind of a snapshot of activity and happenings. Some of its fields represent
+// "moment-in-time" values e.g. BufferSize is the size of the buffer at the moment the Stat is
+// sent. Other fields are cumulative since the last Stat, i.e. ErrorsSinceLastStat.
+type BatchStat struct {
```

Comment:
It seems to me that now that this is all in a package named `batchproducer`, this type can be named just `Stat`. Oh I think I see why you found the name confusing — because it’s a singular name but the struct contains multiple stats — each field is in effect a stat. Yeah… perhaps this should be called something like `StatsSnapshot` or something… I want to try to convey that there will be a bunch of these in succession… they’re kinda like frames. Maybe `StatsFrame`?